autorest/client.go

```diff
@@ -262,3 +311,80 @@ func (c Client) ByInspecting() RespondDecorator {
 	}
 	return c.ResponseInspector
 }
+
+func withLogging() PrepareDecorator {
+	return func(p Preparer) Preparer {
+		return PreparerFunc(func(r *http.Request) (*http.Request, error) {
+			// execute any remaining preparers before logging
+			r, err := p.Prepare(r)
+			if logLevel > logLevelNone {
+				b := &bytes.Buffer{}
+				fmt.Fprintf(b, "(%s) REQUEST: %s %s\n", time.Now().Format(logTimestamp), r.Method, r.URL.String())
+				// dump headers
+				for k, v := range r.Header {
+					if strings.ToLower(k) == "authorization" {
+						v = []string{"**REDACTED**"}
+					}
+					fmt.Fprintf(b, "%s: %s\n", k, strings.Join(v, ","))
+				}
+				ct := r.Header.Get(headerContentType)
+				if logLevel == logLevelFull && r.Body != nil && strings.Index(ct, mimeTypeOctetStream) == -1 {
+					// dump body
+					var body []byte
+					body, err = ioutil.ReadAll(r.Body)
+					if err == nil {
+						fmt.Fprintln(b, string(body))
+						if nc, ok := r.Body.(NopClosingReadSeeker); ok {
+							// rewind to the beginning
+							nc.Seek(0, io.SeekStart)
+						} else {
+							// recreate the body
+							r.Body = NopClosingReadSeeker{bytes.NewReader(body)}
+						}
+					} else {
+						fmt.Fprintf(b, "failed to read body: %v\n", err)
+					}
+				}
+				logFile.WriteString(b.String())
+				logFile.Sync()
+			}
+			return r, err
+		})
+	}
+}
+
+func byLogging() RespondDecorator {
+	return func(r Responder) Responder {
+		return ResponderFunc(func(resp *http.Response) error {
+			var err error
+			if logLevel > logLevelNone {
+				b := &bytes.Buffer{}
+				fmt.Fprintf(b, "(%s) RESPONSE: %d %s\n", time.Now().Format(logTimestamp), resp.StatusCode, resp.Request.URL.String())
+				// dump headers
+				for k, v := range resp.Header {
+					fmt.Fprintf(b, "%s: %s\n", k, strings.Join(v, ","))
+				}
+				ct := resp.Header.Get(headerContentType)
+				if logLevel == logLevelFull && resp.Body != nil && strings.Index(ct, mimeTypeOctetStream) == -1 {
+					// dump body
+					defer resp.Body.Close()
```

Comment:
I think that this will end up calling `Close()` on the wrong Body, as you'll want to close the original after the `ioutil.ReadAll` a few lines below. Using `defer`, Close will get called on the reassigned `NopClosingReadSeeker`.

Reply:
"A deferred function's arguments are evaluated when the defer statement is evaluated."
See https://blog.golang.org/defer-panic-and-recover

Reply:
Ah, my bad! I guess I had this one wrong :)

For anyone else who's curious I wrote up a little example: https://play.golang.org/p/nJD9JoAgDti

Reply:
No worries I did pretty much the exact same exercise then found this link.